File: codechef/August13/seabal/seabal.c
#include<stdio.h>

int fast_int()
{
        int n = 0;
    int negflag=0;
        char c = getchar_unlocked();
    if(c=='-')
        negflag=1;
        while (!('0' <= c && c <= '9'))
        {
                c = getchar_unlocked();
        }
        while ('0' <= c && c <= '9')
        {
                n = (n<<3)+(n<<1) + c - '0';
                c = getchar_unlocked();
        }
    if(negflag==1)
        n*=-1;
        return n;
}
// Building Segment Tree - Begin
// node = 1
// b = first position of our array A[]
// e = last position of our array A[]
// M = tree[] array
// A = a[] array
void createTree(int node, int b, int e, long long SumTree[], int A[])
{
    if (b == e)
    {
        SumTree[node] = A[b];
    }
    else
    {
        //compute the values in the left and right subtrees
        createTree(2 * node, b, (b + e) / 2, SumTree, A);
        createTree(2 * node + 1, (b + e) / 2 + 1, e, SumTree, A);

        //doing stuff with children nodes
        //In this case, we are storing sum of both the children nodes in parent node
        SumTree[node] = SumTree[2*node] + SumTree[2*node + 1];
    }
}

// Building Segment Tree - End
// Updating Tree on changing any node value
// idx = position which is to be updated by value 'val'
void update(int node, int b, int e, long long SumTree[], int idx, int val)
{
    if (b > e || b > idx || e < idx ) return;               // this condition is damn important
    if (b == e) //at leaf node
    {
        SumTree[node] = val;
        return;
    }
    update(node*2, b, (b+e)/2, SumTree, idx, val);
    update(node*2 + 1, (b+e)/2 + 1, e, SumTree, idx, val);

    SumTree[node] = SumTree[2*node] + SumTree[2*node + 1];
}

// goku = range begin
// naruto - range end
long long findRangeSum(int node, int b, int e, long long SumTree[], int goku, int naruto)
{
    if(goku <= b && naruto >= e)     // yeah... this condition is very confusing
    {
        return SumTree[node];
    }
    if(b > naruto || e < goku)      // means you are in the wrong neighbourhood buddy
    {
        return 0;
    }
    return findRangeSum(2*node, b, (b+e)/2, SumTree, goku, naruto) + findRangeSum(2*node + 1, (b+e)/2 + 1, e, SumTree, goku, naruto);
}

int main()
{
    int n,m,i,j;
    n = fast_int();
    m = fast_int();
    int a[n+1];
    for(i=0;i<n;i++)
    {
        a[i] = fast_int();
    }
    long long sumTree[262143 + 1000] = {0};
    createTree(1,0,n-1,sumTree,a);
    long long pairs[m];
    int l[m+1];
    int r[m+1];
    for(i=0;i<m;i++)
    {
        l[i] = fast_int();
        r[i] = fast_int();
        pairs[i] = findRangeSum(1,0,n-1,sumTree,l[i]-1,r[i]-1);
    }
    int k;
    int count;
    int change = 0;

    k = fast_int();
    while(k--)
    {
        int x,y;
        count = 0;
        x = fast_int();
        y = x + change;
        for(j=0;j<m;j++)
        {
            if(y>=l[j] && y<=r[j])
            {
                pairs[j]--;
            }
            if(pairs[j] == 0)
            {
                count++;
            }
        }
        change = count;
        printf("%d\n",count);
    }
    return 0;
}

File: codechef/August13/seabal/seabal.py
from sys import stdin

def createTree(node, b, e, SumTree, MaxIndexTree, A):
    if b == e:
        SumTree[node] = A[b]
        MaxIndexTree[node] = b
    else:
        createTree(2*node, b, (b+e)/2, SumTree, MaxIndexTree, A)
        createTree(2*node + 1, (b+e)/2 + 1, e, SumTree, MaxIndexTree, A)
        SumTree[node] = SumTree[2*node] + SumTree[2*node + 1]
        if A[MaxIndexTree[2*node]] >= A[MaxIndexTree[2*node + 1]]:
            MaxIndexTree[node] = MaxIndexTree[2*node]
        else:
            MaxIndexTree[node] = MaxIndexTree[2*node + 1]

def update(node, b, e, SumTree, MaxIndexTree, idx, val, A):
    if b>e or b>idx or e < idx:
        return
    if b == e:
        SumTree[node] = val
        return
    update(2*node, b, (b+e)/2, SumTree, MaxIndexTree, idx, val, A)
    update(2*node + 1, (b+e)/2 + 1, e, SumTree, MaxIndexTree, idx, val, A)
    SumTree[node] = SumTree[2*node] + SumTree[2*node + 1]
    if A[MaxIndexTree[2*node]] >= A[MaxIndexTree[2*node + 1]]:
        MaxIndexTree[node] = MaxIndexTree[2*node]
    else:
        MaxIndexTree[node] = MaxIndexTree[2*node + 1]

def findRangeSum(node, b, e, SumTree, goku, naruto):
    if goku <= b and naruto >= e:
        return SumTree[node]
    if b > naruto or e < goku:
        return 0
    return findRangeSum(2*node, b, (b+e)/2, SumTree, goku, naruto) + findRangeSum(2*node + 1, (b+e)/2 + 1, e, SumTree, goku, naruto)

def findRangeMaxIndex(node, b, e, MaxIndexTree, goku, naruto, A):
    if goku <= b and naruto >= e:
        return MaxIndexTree[node]
    if b > naruto or e < goku:
        return -1
    q = findRangeMaxIndex(2*node, b, (b+e)/2, MaxIndexTree, goku, naruto, A)
    w = findRangeMaxIndex(2*node + 1, (b+e)/2 + 1, e, MaxIndexTree, goku, naruto, A)
    if A[q] >= A[w]:
        return q
    else:
        return w
    #return findRangeMaxIndex(2*node, b, (b+e)/2, MaxTree, goku, naruto, A) + findRangeSum(2*node + 1, (b+e)/2 + 1, e, MaxTree, goku, naruto, A)

def seabal():
    d = 0
    data = stdin.readlines()
    n,m = map(int,data[d].split())
    d += 1
    a = map(int,data[d].split())
    a.append(-1)
    d += 1
    #print n,m,a
    sumTree = [0]*263143
    maxIndexTree = [0]*263143
    createTree(1,0,n-1,sumTree,maxIndexTree,a)
    clones = [[] for x in xrange(n)]
    l = []
    r = []
    pairs = []
    for i in xrange(m):
        t = map(int,data[d].split())
        d += 1
        l.append(t[0])
        r.append(t[1])
        clones[findRangeMaxIndex(1, 0, n-1, maxIndexTree, l[i]-1, r[i]-1, a)].append(i)
        pairs.append(findRangeSum(1,0,n-1,sumTree,l[i]-1,r[i]-1))
    print clones
    change = 0
    count = 0
    k = int(data[d])
    d += 1
    while k!=0:
        count = 0
        x = int(data[d])
        d += 1
        y = x + change
        for i in xrange(m):
            if y >= l[i] and y <= r[i]:
                pairs[i] -= 1
            if pairs[i] == 0:
                count += 1
        change = count
        print count
        k -= 1
seabal()
